File: IPWorks Auth Samples/OTP Demo/java/otp.java
/*
 * IPWorks Auth 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Auth in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksauth
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.*;
import ipworksauth.*;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class otp extends ConsoleDemo {
    public static void main(String[] args) {
        OTP hotp = new OTP();
        OTP totp = new OTP();
        try {
            String choice="";
            while (true) {
                System.out.println("Choose the OTP algorithm:");
                System.out.println("1) HOTP");
                System.out.println("2) TOTP");
                System.out.println("Q (Quit)");
                choice = prompt("otp", ">");
                if (choice.equalsIgnoreCase("hotp")||choice.equalsIgnoreCase("1")) {
                    String option = "";
                    while (!option.equalsIgnoreCase("q")&&!option.equalsIgnoreCase("quit")) {
                        System.out.println("Please select an operation:");
                        System.out.println("1)  Create Password");
                        System.out.println("2)  Validate Password");
                        System.out.println("Q)  Quit");
                        option = prompt("hotp", ">");
                        hotp.reset();
                        hotp.setPasswordAlgorithm(1);
                        // Initial setup
                        if (option.equals("1")) {
                            hotp.setSecret(prompt("Secret", ":", "ABCDEFGHIJKLMNOP"));
                            hotp.setCounter(Integer.parseInt(prompt("Counter", ":", "1")));
                            hotp.createPassword();
                            System.out.println("Password: " + hotp.getPassword() + "\r\n");
                        } else if (option.equals("2")) {
                            hotp.setSecret(prompt("Secret", ":", "ABCDEFGHIJKLMNOP"));
                            hotp.setCounter(Integer.parseInt(prompt("Counter", ":", "1")));
                            hotp.setPassword(prompt("Password", ":"));
                            if (hotp.validatePassword()) {
                                System.out.println("VALID!\r\n");
                            } else {
                                System.out.println("INVALID!\r\n");
                            }
                        } else if (!option.equalsIgnoreCase("q")&&!option.equalsIgnoreCase("quit")) {
                            System.out.println("Error: Input not recognized.");
                        } else { continue;
                        }
                    }
                } else if (choice.equalsIgnoreCase("totp")||choice.equalsIgnoreCase("2")) {
                    String option = "";
                    while (!option.equalsIgnoreCase("q")&&!option.equalsIgnoreCase("quit")) {
                        System.out.println("Please select an operation:");
                        System.out.println("1)  Create Password");
                        System.out.println("2)  Validate Password");
                        System.out.println("Q)  Quit");
                        option = prompt("totp", ">");
                        totp.setPasswordAlgorithm(0); //also by default
                        // Initial setup
                        if (option.equals("1")) {
                            totp.setSecret(prompt("Secret", ":", "ABCDEFGHIJKLMNOP"));
                            totp.setTimeStep(Integer.parseInt(prompt("TimeStep", ":", "30")));
                            totp.createPassword();
                            System.out.println("Validity Time: " + totp.getValidityTime() + "\r\n");
                            System.out.println("Password: " + totp.getPassword());
                        } else if (option.equals("2")) {
                            System.out.println("Validity Time: " + totp.getValidityTime() + "\r\n");
                            totp.setSecret(prompt("Secret", ":", "ABCDEFGHIJKLMNOP"));
                            totp.setTimeStep(Integer.parseInt(prompt("TimeStep", ":", "30")));
                            totp.setPassword(prompt("Password", ":"));
                            if (totp.validatePassword()) {
                                System.out.println("VALID!\r\n");
                            } else {
                                System.out.println("INVALID!\r\n");
                            }
                        } else if (!option.equalsIgnoreCase("q")&&!option.equalsIgnoreCase("quit")) {
                            System.out.println("Error: Input not recognized.");
                        } else { continue;
                        }
                    }
                } else if(choice.equalsIgnoreCase("q")||choice.equalsIgnoreCase("quit")){
                    break;
                }else { System.out.println("Error: Input not recognized.");
                }
            }
        } catch(IPWorksAuthException ex){
            System.out.println("IPWorksAuth exception thrown: " + ex.getCode()
                    + " [" + ex.getMessage() + "].");
        } catch(Exception ex){
            System.out.println(ex.getMessage());
        }
    }
}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksAuthException) {
      System.out.print(" (" + ((IPWorksAuthException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks Auth Samples/OTP Demo/python/otp.py
# 
# IPWorks Auth 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks Auth in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworksauth
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworksauth import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


import sys
import string
from ipworksauth import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] == None:
    args[index] = input(prompt)

def fireError(e):
  print(e.message)

hotp = OTP()
totp = OTP()
hotp.on_error = fireError
totp.on_error = fireError

try:
	option = ""
	choice = ""
	while choice != "q" and choice!= "quit":
		print("Please select the algorithm:")
		print("1) HOTP")
		print("2) TOTP")
		print("q) Quit")
		choice = input("Option > ").lower()

		if choice == "1" or choice=="hotp":				
			while True:
				print("Please select the option:")
				print("1) Create Password")
				print("2) Validate Password")
				print("q) Quit")
				option = input("hotp > ").lower()
				hotp.reset()
				hotp.set_password_algorithm(1)
				if option=="1":
					print("Create password!")
					secret = input("Secret: ")
					if secret == "":
						secret = "ABCDEFGHIJKLMNOP"
					hotp.set_secret(secret)
					try:
						counter = input("Counter: ")
						if counter == "":
							counter = 1
						hotp.set_counter(int(counter))
					except ValueError:
						print("Error: Input not recognized")
						continue
					hotp.create_password()
					print("Password: " + hotp.get_password() + "\r\n")	
					continue
				if option=="2":			
					print("Validate password!")
					secret = input("Secret: ")
					if secret == "":
						secret = "ABCDEFGHIJKLMNOP"
					hotp.set_secret(secret)					
					try:
						counter = input("Counter: ")
						if counter == "":
							counter = 1
						hotp.set_counter(int(counter))
					except ValueError:
						print("Error: Input not recognized")
						continue
					password = input("Password: ")			
					hotp.set_password(str(password))					
					if hotp.validate_password():
						print("VALID")
						continue
					else:
						print("INVALID")
						continue
				elif option=="q"or option=="quit":
					break
				else:
					print("Error: Input not recognized")
					continue
		elif choice == "2" or choice=="totp":			
			while True:		
				print("Please select the option:")
				print("1) Create Password")
				print("2) Validate Password")
				print("q) Quit")
				option = input("totp > ").lower()		
				totp.set_password_algorithm(0) 		#default algorthim 
				if option=="1":
					print("Create password!")
					secret = input("Secret: ")
					if secret == "":
						secret = "ABCDEFGHIJKLMNOP"
					totp.set_secret(secret)									
					try:
						timestep = input("TimeStep: ")				
						if timestep=="":
							timestep=totp.get_time_step() # default 30 seconds
						totp.set_time_step(int(timestep))	
					except ValueError:
						print("Error: Input not recognized")
						continue
					totp.create_password()
					print("\nPassword: " + totp.get_password() + "\r")	
					print("Validity time: " + str(totp.get_validity_time())+"\n")	
					continue
				elif option=="2":
					print("Validate password!")
					print("Validity time: " + str(totp.get_validity_time()))
					secret = input("Secret: ")
					if secret == "":
						secret = "ABCDEFGHIJKLMNOP"
					totp.set_secret(secret)	
					try:
						timestep = input("TimeStep: ")				
						if timestep=="":
							timestep=totp.get_time_step()
						totp.set_time_step(int(timestep))	
					except ValueError:
						print("Error: Input not recognized")
						continue
					password = input("Password: ")			
					totp.set_password(str(password))
					if totp.validate_password():
						print("VALID")						
						continue
					else:
						print("INVALID")
						continue
				elif option=="q"or option=="quit":
					break
				else:
					print("Error: Input not recognized")
					continue			
		elif choice != "q" and choice != "quit":
			print("Error: Input not recognized\n")
			continue
		else:
			break
except IPWorksAuthError as e:
  print("ERROR %s" %e.message)

